### PR TITLE
Add new input module

### PR DESCRIPTION
**Pull request by toast003**

### CODE REVIEW

**Comment by toast003** [comment]
Had a bit of a brain fart there, sorry for the force pushing

**Comment by toast003** [comment]
@magnouvean I changed the write_config script a bit, since doing base conversion inside nix seems too complicated.
I am very new at python, and although my code seems to work just fine I would appreciate if you could take a look at it

**Comment by toast003** [comment]
Also I don't think the check failing was due to something I did, not sure why
EDIT: nvm it is, I'll fix it later

File: script/write_config.py [hunk 1]
```diff
@@ -76,6 +77,21 @@ def escape(s: str) -> str:
             s[i] = "\\s"
     return "".join(s)
 
+# Some values need to be "transformed", such as being converted from hex to decimal
+def transformValues(dOrig):
+    d = dOrig
```
Comment (magnouvean) [inline script/write_config.py]: Not sure if this is needed, since this won't copy anything, it will just be a pointer to `dOrig`. You can use `dOrig.copy()` if this is wanted (which I think makes sense in this case).

File: script/write_config.py [hunk 2]
```diff
@@ -76,6 +77,21 @@ def escape(s: str) -> str:
...
+    cPath = os.path.expanduser("~/.config/kcminputrc")
+    if cPath in d:
+        for itemStr in list(d[cPath]):
+            if itemStr.startswith("Libinput/"):
+                item = d[cPath].pop(itemStr)
+                itemStrList = itemStr.split("/")
+                itemStrList[1] = str(int(itemStrList[1],16))
+                itemStrList[2] = str(int(itemStrList[2],16))
+                itemStr = "/".join(itemStrList)
+                d[cPath][itemStr] = item
```
Comment (magnouvean) [inline script/write_config.py]: In python it is standard practice to use snake_case, not camelCase (unlike nix, java, C# etc). I think this should be the case here as well.

File: script/write_config.py [hunk 3]
```diff
@@ -329,7 +345,8 @@ def main():
 
     reset_files = set(sys.argv[2].split(" ")) if sys.argv[2] != "" else set()
     immutable_by_default = bool(sys.argv[3])
-    d = json.loads(json_str)
+    dRaw = json.loads(json_str)
+    d = transformValues(dRaw)
```
Comment (magnouvean) [inline script/write_config.py]: snake_case is better here as well

File: modules/input.nix [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    };
+    productId = mkOption {
+      type = with types; nullOr str;
+      default = null;
+      example = "21128";
+      description = ''
+        The product ID of the trackpad
+        You can find it out running cat /proc/bus/input/devices | grep -i touchpad
+      '';
+    };
```
Comment (magnouvean) [inline modules/input.nix]: These three have allowed values as `null` but if the touchpad is enabled and these are not set I'm pretty sure it will crash. This should probably be checked in an assertion so we could give out a proper error message.

File: modules/input.nix [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            LayoutList.value = strings.concatStringsSep "," cfg.input.keyboard.layouts;
+          };
+        }
+      )
+    ]
+  );
+
+  # Trackpad options
+
+  options.programs.plasma.input.touchpad = {
```
Comment (magnouvean) [inline modules/input.nix]: The options here will write to the same group as mentioned in #47, am I right? If so we should probably add `DisableWhileTyping`, `NaturalScroll` and `TapToClick` as options which would close that issue.

**Review by magnouvean — COMMENTED** [review]
I can't really test much atm, but looking at the code in general this looks fine after the minor suggested tweaks :)

**Comment by magnouvean** [comment]
Also https://github.com/nix-community/plasma-manager/blob/d12cb0ed42a81ff84330b5cc94bff6abfeeb6da8/README.md?plain=1#L53 should probably be edited or removed when we're done with this pr.

File: script/write_config.py (hunk 2 above)
Comment (toast003) [reply script/write_config.py]: This is my first real python code and I also happen to come from a java/c# background so I didn't know this, will change now

**Comment by toast003** [comment]
I will when I finish

File: modules/input.nix (hunk 5 above)
Comment (toast003) [reply modules/input.nix]: Yes!
I was going to add those first, but I thought that making the user look for the IDs of the trackpad, and then turning them to decimal was a bit too dumb. 

Already took care of the hexadecimal to decimal part here, but I'll do the "looking for the IDs" part in a separate pr, since thinking about how to do that is what burnt me out, and I don't want this pr to be stalled for another month :(

Comment (magnouvean) [reply modules/input.nix]: No worries, I think just getting this into `plasma-manager` gets us a long way.
Looking for the IDs isn't necessarily going to be that easy, and would, if done at activation-time not be reproducible either (since the output would depend on the devices attached when the activation is applied). I think when doing that we could maybe consider just altering the config-file generated automatically by kde, which contains these ids aleady, in a startup-script, as that would be declarative, and maybe easier to do. I would maybe require a reload of some kde component, but `plasma-manager` should support that already.

Anyway that will be in the future, there's no need to think about this yet (and I think that it isn't necessarily a must for this module anyway).

**Comment by magnouvean** [comment]
I'll try to add support for configuring multiple touchpads and add some assertions today. I think I'll just commit directly here if that's OK :)

**Comment by magnouvean** [comment]
Imo it's more or less good enough to merge in the current state. I haven't found any issues (though I also haven't done enough testing to confirm a lot of edge-cases). There are obviously some tweaks as mentioned above such as changing the README and maybe also adding some examples.

**Comment by HeitorAugustoLN** [comment]
@toast003 I sent you a pull request that adds configurations for mouses in the input module.

**Comment by toast003** [comment]
You guys did all of my work :sweat_smile: 

@HeitorAugustoLN I'll take a look at it rn

**Comment by toast003** [comment]
All seems good to me!

I'll change the readme and after that I think we can merge the PR

File: modules/input.nix [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+          Enables or disables the mouse
+        '';
+      };
+      name = mkOption {
+        type = types.str;
+        default = null;
+        example = "Logitech G403 HERO Gaming Mouse";
+        description = ''
+          The name of the mouse.
+          You can find it out running cat /proc/bus/input/devices | grep -B 1 -i mouse
```
Comment (magnouvean) [inline modules/input.nix]: This doesn't work for me because my mouse doesn't have anything with mouse in the name. `grep '^H:.*mouse' -B 5 /proc/bus/input/devices` seems to work for me though. Seemingly this could maybe also be considered for the touchpad as well since I'm not sure all touchpads include `touchpad` in the name, but they should have `mouse*` in the handlers section

**Review by magnouvean — COMMENTED** [review]
I'll test more right away

File: modules/input.nix (hunk 6 above)
Comment (HeitorAugustoLN) [reply modules/input.nix]: image
It works for me. But it is probably not a general rule.

Comment (magnouvean) [reply modules/input.nix]: Yeah that works because the mouse has "mouse" in the name, but for mine (a logitech mouse as well) it doesn't have this. In your case though the command I provided should work as well though?

File: modules/input.nix [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+    let
+      mouseName = mouse.name;
+      mouseVendor = mouse.vendorId;
+      mouseProduct = mouse.productId;
+    in
+    {
+      "Libinput/${mouseVendor}/${mouseProduct}/${mouseName}" = {
+        Enabled = mouse.enable;
+        LeftHanded = mouse.leftHanded;
+        MiddleMouseEmulation = mouse.middleMouseEmulation;
```
Comment (magnouvean) [inline modules/input.nix:282]: I also think it should be `MiddleButtonEmulation` here

**Comment by magnouvean** [comment]
But yeah other than the two minor things above this work great so we can probably merge after they are fixed :)

File: modules/input.nix (hunk 6 above)
Comment (HeitorAugustoLN) [reply modules/input.nix]: Your command works too. 
image

File: modules/input.nix (hunk 7 above)
Comment (HeitorAugustoLN) [reply modules/input.nix:282]: You're right, I mixed up the names

Comment (HeitorAugustoLN) [reply modules/input.nix:282]: @toast003 could you fix this?

File: modules/input.nix (hunk 6 above)
Comment (magnouvean) [reply modules/input.nix]: Great. My suggestion to solve this would be to replace the old commands (in the three descriptions for both mouse and touchpad) with `grep '^H:.*mouse' -B 5 /proc/bus/input/devices`. We might be able to get away with the old command for touchpads, but afaik it's no requirement for those to have `touchpad` in the name either so I think the new command is still better for touchpads as well.

Comment (toast003) [reply modules/input.nix]: For trackpads you can just grep  `/proc/bus/input/devices` since most trackpads are i2c devices and (afaik) i2c devices don't have names, so trackpads always have trackpad in the name.
For mice we could just check `lsusb`
image

File: modules/input.nix (hunk 7 above)
Comment (toast003) [reply modules/input.nix:282]: Done

Comment (HeitorAugustoLN) [reply modules/input.nix:282]: Thanks!

File: modules/input.nix (hunk 6 above)
Comment (magnouvean) [reply modules/input.nix]: On my end when running the grep-command as it is now it matches `ETPS/2 Elantech Touchpad` which is in the name section of the touchpad in `/proc/bus/input/devices`, and if it was called something else it may not have matched. With that said I'm not sure if it's a real issue as most touchpads probably contain "touchpad" in some form in the name (unless it's called something with trackpad or something).
As for `lsusb` there I can see two potential problems, the first being that `lsusb` isn't installed on most nixos machines I don't think, so it's less accessible than just using `/proc/bus/input/devices`, and second it may not support bluetooth mice (I don't know if it does). The command provided above does just uses `/proc/bus/input/devices` and would work for both touchpads and mice

Comment (toast003) [reply modules/input.nix]: While I don't think your first problem is that big of a deal (you can use comma or `nix run`), I do agree with the second.
lsusb only lists usb devices, so it won't catch anything else. For now I would say to leave both, but we could consider making a python script that figures it out for you

Comment (magnouvean) [reply modules/input.nix]: Yeah figuring this out automatically could be done through some startup script or something and is a task for later. My point was that the descriptions for the options should be correct/helpful. For that the grep command above would suffice for now I reckon, or alternatively we could just add "see /proc/bus/input/devices for the values" or something. A.t.m. the grep commands in the descriptions for the mice doesn't work and the grep commands for the touchpad only works for some hardware (only when the name contains touchpad), and I think this should be correct before we merge into `trunk`.

Comment (HeitorAugustoLN) [reply modules/input.nix]: I think a more general description like `see /proc/bus/input/devices for the values` would be better

File: script/write_config.py [hunk 8]
```diff
@@ -76,6 +77,21 @@ def escape(s: str) -> str:
...
+    if c_path in d:
+        for item_str in list(d[c_path]):
+            if item_str.startswith("Libinput/"):
+                item = d[c_path].pop(item_str)
+                item_str_list = item_str.split("/")
+                item_str_list[1] = str(int(item_str_list[1],16))
+                item_str_list[2] = str(int(item_str_list[2],16))
+                item_str = "/".join(item_str_list)
+                d[c_path][item_str] = item
+    return d;
```
Comment (apfelkuchen6) [inline script/write_config.py:94]: I don't really like this because it changes the interface of the `configFile` option in an unexpected manner. It's fine if we want to have the options of the input module expect the ids in hex, but then the conversion should also be done in the module (possibly with something like [this](https://github.com/Misterio77/nix-colors/blob/b92df8f5eb1fa20d8e09810c03c9dc0d94ef2820/lib/core/conversions.nix#L72-L94)).

This change broke my config, which included `programs.plasma.configFile.kcminputrc."Libinput/1390/269/ELECOM TrackBall Mouse HUGE TrackBall".ScrollMethod.value =  4;`. Now the ids are silently mangled.

Comment (magnouvean) [reply script/write_config.py:94]: I agree there. It should be fairly straight forward to do so as well. I'll try to do the hex-conversion in nix instead and open a PR about it.

Comment (HeitorAugustoLN) [reply script/write_config.py:94]: >  but then the conversion should also be done in the module

This would be indeed better.

Comment (magnouvean) [reply script/write_config.py:94]: Should be fixed now after #295.
